napari/utils/events/dataclass.py

```diff
@@ -181,12 +181,15 @@ def set_with_events(self: C, name: str, value: Any) -> None:
     # if different we emit the event with new value
     after = getattr(self, name)
     if isinstance(before, np.ndarray) or isinstance(after, np.ndarray):
-        different = np.any(before != after)
+        different = (
+            np.any(before != after) if before.size == after.size else True
```

Comment:
this will definitely save time when the arrays are different sizes... unfortunately, we now can't use `or` in the conditional, or we might get an `AttributeError` when accessing `.size`, yet, will still have to know if _either_ of them is an array :/

Also, let's use `np.array_equal` rather than `np.any` to avoid the deprecation warnings that @sofroniewn saw.  

we also need to know if they are dask arrays, and not just numpy arrays (but bail out short of comparison, and just assume they are different)... otherwise `bool(before != after)` will still fail.

I did some tests and it definitely does help to check if they're arrays first (as opposed to simply using `np.array_equal` directly).  Here's the fastest, most general solution I could come up with (if not the prettiest):
```python
def is_different(a, b):
    if hasattr(a, 'size'):
        if hasattr(b, 'size'):
            return not np.array_equal(a, b) if a.size == b.size else True
        return True
    if hasattr(b, 'size'):
        return True
    return a != b
```

that will raise a FutureWarning if either is a dask array... so feel free to keep modifying and playing around with the idea.

Reply:
I  add this because I got FutureWarnning for compare array of different size when running test set. 
Now I see that we should add some test to pass tuple, list and dask array. And fix it.

Reply:
@tlambert03 @sofroniewn I try to think a little about this. And I'm not sure if the signal should not be emitted if there are two different representations of the same array.

If value is `[1, 2, 3]` and senewtted value is `numpy.array([1, 2, 3])` It may be new semanic value (some object could treat list and numpy array different?

Reply:
yeah, I was thinking about that too.  Since this is a very low-level object that will be used in many different ways, I think we should err on the side of emission (and fast evaluation), rather than try to get clever about avoiding signal emission for extremely similar (but technically different) objects.  There are [other ways](https://github.com/napari/napari/blob/2508efc85a1141cf6ac9d422f4dc9ce6a69d6c1f/docs/source/explanations/dataclass.md#gettersetter-side-effects--conditional-event-logic) for users to prevent event emission if it's really important.

Reply:
The current comparison also will fail when having dict with nested numpy arrays. 
I found such package: https://pypi.org/project/deepdiff/

What did you think about adding this dependency? Something like:
```
def compare(v1, v2):
    if isinstance(v1, da.core.Array) or isinstance(v2, da.core.Array):
        return False
    diff = DeepDiff(v1, v2, ignore_numeric_type_changes=True, max_diffs=1)
    return diff == {}
```

Reply:
as mentioned, I'm hesitant to add nearly _any_ overhead in this particular function (since it can be added in `_on_<name>_set` if careful diffing is required).  But, if you want to do some tests with `%%timeit` and can demonstrate that it doesn't get any slower while also meeting more edge cases, i guess we could consider the dependency

Reply:
(I'm also fine just having a try/catch and saying "yes they're different" if the comparison fails)

Reply:
Ready